### PR TITLE
Pr/ail 95

### PR DESCRIPTION
**Pull request by mwever**
Automated feature engineering for relational data.

### CODE REVIEW

File: JAICore/jaicore-basic/src/jaicore/basic/SQLAdapter.java [hunk 1]
```diff
@@ -0,0 +1,295 @@
+package jaicore.basic;
+
+import java.io.Serializable;
```
Comment (fmohr) [inline JAICore/jaicore-basic/src/jaicore/basic/SQLAdapter.java]: I think that this file is not in a valid source folder

File: JAICore/jaicore-experiments/src/jaicore/experiments/ExperimentRunner.java [hunk 2]
```diff
@@ -0,0 +1,351 @@
+package jaicore.experiments;
```
Comment (fmohr) [inline JAICore/jaicore-experiments/src/jaicore/experiments/ExperimentRunner.java]: I think that this file is not in a valid source folder

File: build.gradle [hunk 3]
```diff
@@ -182,6 +182,7 @@ allprojects {
 
 dependencies{
 	compile project(":hasco")
```
Comment (fmohr) [inline build.gradle]: I am a bit confused about this dependency section. Why do we have these dependencies defined globally?

File: softwareconfiguration/autofe-reldata/conf/rankingExpConf.txt [hunk 4]
```diff
@@ -0,0 +1,14 @@
+mem.max = 16392
+cpu.max = 6
+
+db.host = localhost
```
Comment (fmohr) [inline softwareconfiguration/autofe-reldata/conf/rankingExpConf.txt]: Credentials (even dummies) should not be committed

File: softwareconfiguration/autofe-reldata/src/main/java/autofe/db/configuration/DatabaseAutoFeConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
+package autofe.db.configuration;
+
+public class DatabaseAutoFeConfiguration {
```
Comment (fmohr) [inline softwareconfiguration/autofe-reldata/src/main/java/autofe/db/configuration/DatabaseAutoFeConfiguration.java]: I feel that this should be a configuration interface instead.

File: softwareconfiguration/autofe-reldata/src/main/java/autofe/db/model/relation/BackwardRelationship.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+	}
+
+	public BackwardRelationship(String fromTableName, String toTableName, String commonAttributeName) {
+		super(fromTableName, toTableName, commonAttributeName);
+	}
+
+	@Override
+	public String toString() {
+		return "BackwardRelationship [fromTableName=" + fromTableName + ", toTableName=" + toTableName
+				+ ", commonAttributeName=" + commonAttributeName + "]";
```
Comment (fmohr) [inline softwareconfiguration/autofe-reldata/src/main/java/autofe/db/model/relation/BackwardRelationship.java]: I think that code formatting here differs from our (line break)

File: softwareconfiguration/autofe-reldata/src/main/java/autofe/db/search/DatasetEvaluationFailedException.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package autofe.db.search;
+
+public class DatasetEvaluationFailedException extends RuntimeException {
+
```
Comment (fmohr) [inline softwareconfiguration/autofe-reldata/src/main/java/autofe/db/search/DatasetEvaluationFailedException.java]: I think that this kind of exceptions should be checked (not inherit from RuntimeException)

File: softwareconfiguration/autofe-reldata/src/main/java/autofe/db/search/NoSolutionFromRandomCompletionException.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package autofe.db.search;
+
+public class NoSolutionFromRandomCompletionException extends RuntimeException {
+
```
Comment (fmohr) [inline softwareconfiguration/autofe-reldata/src/main/java/autofe/db/search/NoSolutionFromRandomCompletionException.java]: Should be a checked Exception I suppose

File: softwareconfiguration/autofe-reldata/src/main/java/autofe/db/sql/RetrieveInstancesFromDatabaseFailedException.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package autofe.db.sql;
+
+public class RetrieveInstancesFromDatabaseFailedException extends RuntimeException {
```
Comment (fmohr) [inline softwareconfiguration/autofe-reldata/src/main/java/autofe/db/sql/RetrieveInstancesFromDatabaseFailedException.java]: Make this a checked exception?

File: softwareconfiguration/autofe-reldata/src/main/java/autofe/db/sql/UnsupportedAttributeTypeException.java [hunk 10]
```diff
@@ -0,0 +1,18 @@
+package autofe.db.sql;
+
+public class UnsupportedAttributeTypeException extends RuntimeException {
+
```
Comment (fmohr) [inline softwareconfiguration/autofe-reldata/src/main/java/autofe/db/sql/UnsupportedAttributeTypeException.java]: Checked Exception?

File: softwareconfiguration/autofe-reldata/src/main/java/autofe/db/util/Tuple.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
+package autofe.db.util;
+
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+
+public class Tuple<T, U> {
```
Comment (fmohr) [inline softwareconfiguration/autofe-reldata/src/main/java/autofe/db/util/Tuple.java]: I believe that this class is a duplicate of our `Pair` class

File: softwareconfiguration/autofe-reldata/src/main/java/autofe/processor/NoSolutionFoundException.java [hunk 12]
```diff
@@ -0,0 +1,18 @@
+package autofe.processor;
+
+public class NoSolutionFoundException extends RuntimeException {
```
Comment (fmohr) [inline softwareconfiguration/autofe-reldata/src/main/java/autofe/processor/NoSolutionFoundException.java]: Checked Exception

File: softwareconfiguration/autofe-reldata/src/main/java/autofe/util/InvalidEvaluationFunctionException.java [hunk 13]
```diff
@@ -0,0 +1,18 @@
+package autofe.util;
+
+public class InvalidEvaluationFunctionException extends RuntimeException {
```
Comment (fmohr) [inline softwareconfiguration/autofe-reldata/src/main/java/autofe/util/InvalidEvaluationFunctionException.java]: Checked Exception

File: softwareconfiguration/autofe-reldata/src/test/java/autofe/db/test/SearchTest.java [hunk 14]
```diff
@@ -0,0 +1,349 @@
...
+		DatabaseNodeEvaluator evaluator = new DatabaseNodeEvaluator(generator);
+
+		GraphSearchWithSubpathEvaluationsInput<DatabaseNode, String, Double> tree = new GraphSearchWithSubpathEvaluationsInput<>(
+				generator, evaluator);
+
+		BestFirst<GraphSearchWithSubpathEvaluationsInput<DatabaseNode, String, Double>, DatabaseNode, String, Double> search = new BestFirst<>(
+				tree);
+		search.setTimeoutForComputationOfF(60000, node -> 100.0);
+
+		// TODO: Add graph visualizer
```
Comment (fmohr) [inline softwareconfiguration/autofe-reldata/src/test/java/autofe/db/test/SearchTest.java]: commented code

File: build.gradle (hunk 3 above)
Comment (mwever) [reply build.gradle]: it is not defined globally but for the super-project AILibs which consists of its sub-projects.

File: softwareconfiguration/autofe-reldata/src/main/java/autofe/db/model/relation/BackwardRelationship.java (hunk 6 above)
Comment (mwever) [reply softwareconfiguration/autofe-reldata/src/main/java/autofe/db/model/relation/BackwardRelationship.java]: Resolved

File: softwareconfiguration/autofe-reldata/src/main/java/autofe/db/search/DatasetEvaluationFailedException.java (hunk 7 above)
Comment (mwever) [reply softwareconfiguration/autofe-reldata/src/main/java/autofe/db/search/DatasetEvaluationFailedException.java]: Resolved

File: softwareconfiguration/autofe-reldata/src/main/java/autofe/db/sql/RetrieveInstancesFromDatabaseFailedException.java (hunk 9 above)
Comment (mwever) [reply softwareconfiguration/autofe-reldata/src/main/java/autofe/db/sql/RetrieveInstancesFromDatabaseFailedException.java]: Resolved, now it is a checked exception.

File: softwareconfiguration/autofe-reldata/src/main/java/autofe/db/search/NoSolutionFromRandomCompletionException.java (hunk 8 above)
Comment (mwever) [reply softwareconfiguration/autofe-reldata/src/main/java/autofe/db/search/NoSolutionFromRandomCompletionException.java]: Resolved, now it is a checked exception.

**Comment by mwever** [comment]
Requested changes have been resolved (except for javadoc, this is postponed to a master's thesis).

**Pull request merged**